include/nbl/asset/ISkeleton.h

```diff
@@ -34,59 +38,68 @@ namespace asset
 				return invalid_joint_id;
 			}
 
-			inline uint32_t getJointCount() const
+			inline joint_id_t getJointCount() const
 			{
-				return jointCount;
+				return m_jointCount;
 			}
 
-			inline SBufferBinding<BufferType>& getParentJointIDBinding()
-			{
-				return m_parentJointIDs;
-			}
-			inline const SBufferBinding<BufferType>& getParentJointIDBinding() const
+			inline const SBufferBinding<const BufferType>& getParentJointIDBinding() const
 			{
-				return m_parentJointIDs;
+				return reinterpret_cast<const SBufferBinding<const BufferType>*>(m_parentJointIDs);
 			}
 
-			inline SBufferBinding<BufferType>& getInverseBindPosesBinding()
-			{
-				return m_inverseBindPoses;
-			}
-			inline const SBufferBinding<BufferType>& getInverseBindPosesBinding() const
+			inline const SBufferBinding<const BufferType>& getDefaultTransforms() const
 			{
-				return m_inverseBindPoses;
+				return reinterpret_cast<const SBufferBinding<const BufferType>*>(m_defaultTransforms);
 			}
 
 
 		protected:
-			ISkeleton(SBufferBinding<BufferType>&& _parentJointIDsBinding, SBufferBinding<BufferType>&& _inverseBindPosesBinding)
-				:	m_nameToJointID(), m_stringPoolSize(0ull), m_stringPool(nullptr), m_jointCount(0u),
-					m_parentJointIDs(std::move(_parentJointIDsBinding)), m_inverseBindPoses(std::move(_inverseBindPosesBinding))
+			ISkeleton(SBufferBinding<BufferType>&& _parentJointIDsBinding, SBufferBinding<BufferType>&& _defaultTransforms, joint_id_t _jointCount = 0u)
+				:	m_nameToJointID(), m_stringPoolSize(0ull), m_stringPool(nullptr), m_jointCount(_jointCount),
+					m_parentJointIDs(std::move(_parentJointIDsBinding)), m_defaultTransforms(std::move(_defaultTransforms))
 			{
+				if (m_jointCount==0u)
+					return;
+
+				assert(m_parentJointIDs.buffer->getSize()>=m_parentJointIDs.offset+sizeof(joint_id_t)*m_jointCount);
+				assert(m_defaultTransforms.buffer->getSize()>=m_defaultTransforms.offset+sizeof(core::matrix3x4SIMD)*m_jointCount);
 			}
 			virtual ~ISkeleton()
 			{
-				_NBL_DELETE_ARRAY(m_stringPool,m_stringPoolSize);
+				clearNames();
+			}
+
+			// map must contain one `const char*` per bone
+			template<class Comparator>
+			inline void setJointNames(const core::map<const char*,joint_id_t,Comparator>& nameToJointIDMap)
+			{
+				clearNames();
+
+				// size the pool
+				for (const auto& mapping : nameToJointIDMap)
+					reserveName(mapping.first);
+				
+				// useless names
+				if (m_stringPoolSize==0ull)
+					return;
+
+				m_stringPool = _NBL_NEW_ARRAY(char,m_stringPoolSize);
```

Comment:
why not smart refctd array?

Reply:
no need, plus it becomes the same thing anyway behind the scenes

Reply:
also I dont plan on sharing the array, and thats the use case for `smart_refctd_array`